Generates data fetcher interface for query types

zaenk opened the pull request:
The Java generator has an option to generate example data fetchers that can be copied into the codebase and extended as needed, but the kotlin generator has no such option.

This PR implements example data fetcher generator for kotlin with a twist: instead of generating a a class with TODO bodies, it generates an interface, for now.

This is the first in a series of PRs that would add an option similar to [OpenAPI kotlin-spring generator](https://openapi-generator.tech/docs/generators/kotlin-spring/) `interfaceOnly` option.

The final solution would enable the development of schemas in a separate repository. Automation would use this generator to generate data classes for types and interfaces for data fetchers and mutations that server component must implement. The generated sources would be packaged and published as a jar.

ilterpehlivan commented:
Hi,
This PR would be answering my question here ? https://github.com/Netflix/dgs-framework/discussions/308#discussioncomment-1754302 <br>

Thanks

zaenk commented:
@ilterpehlivan yes, essentially it does, original discussion: https://github.com/Netflix/dgs-framework/discussions/234

zaenk commented:
rebased, tests are now passing

zaenk commented:
Many things changed since I opened this MR. Should this behaviour be also added to the `config.generateKotlinNullableClasses == true` code branch?

zaenk commented:
TODO: resolve issues with the generated annotation

```kotlin
public interface ShowsDatafetcher {
  @DgsData(
    parentType,
    field,
  )
  public fun getShows(titleFilter: String?, dataFetchingEnvironment: DataFetchingEnvironment):
      List<Show?>?
}

```

zaenk commented:
resolved annotation generation issue, also added DgsComponent annotation to this interface

```kotlin
@DgsComponent
public interface ShowsDatafetcher {
  @DgsData(
    parentType = DgsConstants.QUERY.TYPE_NAME,
    field = DgsConstants.QUERY.Shows,
  )
  public fun getShows(titleFilter: String?, dataFetchingEnvironment: DataFetchingEnvironment):
      List<Show?>?
}

```

zaenk commented:
added one more thing, again, oversight from my part: input argument annotation

```kotlin
@DgsComponent
public interface ShowsDatafetcher {
  @DgsData(
    parentType = DgsConstants.QUERY.TYPE_NAME,
    field = DgsConstants.QUERY.Shows,
  )
  public fun getShows(@InputArgument("titleFilter") titleFilter: String?,
      dataFetchingEnvironment: DataFetchingEnvironment): List<Show?>?
}
```

zaenk commented:
changed datafetcher postfix to query, also added generator for mutation and subscription

```kotlin

@DgsComponent
public interface ShowsQuery {
  @DgsData(
    parentType = DgsConstants.QUERY.TYPE_NAME,
    field = DgsConstants.QUERY.Shows,
  )
  public fun shows(@InputArgument(DgsConstants.QUERY.SHOWS_INPUT_ARGUMENT.TitleFilter)
      titleFilter: String?, dataFetchingEnvironment: DataFetchingEnvironment): List<Show?>?
}


@DgsComponent
public interface AddArtworkMutation {
  @DgsData(
    parentType = DgsConstants.MUTATION.TYPE_NAME,
    field = DgsConstants.MUTATION.AddArtwork,
  )
  public fun addArtwork(
    @InputArgument(DgsConstants.MUTATION.ADDARTWORK_INPUT_ARGUMENT.ShowId) showId: Int,
    @InputArgument(DgsConstants.MUTATION.ADDARTWORK_INPUT_ARGUMENT.Upload) upload: MultipartFile,
    dataFetchingEnvironment: DataFetchingEnvironment,
  ): List<Image?>
}

@DgsComponent
public interface AddReviewMutation {
  @DgsData(
    parentType = DgsConstants.MUTATION.TYPE_NAME,
    field = DgsConstants.MUTATION.AddReview,
  )
  public fun addReview(@InputArgument(DgsConstants.MUTATION.ADDREVIEW_INPUT_ARGUMENT.Review)
      review: SubmittedReview?, dataFetchingEnvironment: DataFetchingEnvironment): List<Review?>?
}


@DgsComponent
public interface ReviewAddedSubscription {
  @DgsData(
    parentType = DgsConstants.SUBSCRIPTION.TYPE_NAME,
    field = DgsConstants.SUBSCRIPTION.ReviewAdded,
  )
  public fun reviewAdded(@InputArgument(DgsConstants.SUBSCRIPTION.REVIEWADDED_INPUT_ARGUMENT.ShowId)
      showId: Int, dataFetchingEnvironment: DataFetchingEnvironment): Publisher<Review?>
}

```

zaenk commented:
also addedd datafetcher interface generation to kotlin2 code branch

it would generate the same interface files as the other branch - the difference in this branch is that instead of generating non-nullable fields on data classes, it generates a "supplier". the reason for this is when the client doesn't requests a non-nullable field, the server would still invoke the data fetcher for it, otherwise it would run into a initialization filter. this behaviour doesn't affect the data fetcher interfaces

zaenk commented:
@paulbakker @iuliiasobolevska @srinivasankavitha this feature request adds new functionality without modifying existing behaviour. wdyt, is it something that you can merge into mainline? Short description of the feature below:

I've added a new configuration option to the gradle plugin, `generateDataFetcherInterfaces` - this will generate SAM for each property a Mutation, Query or Subscription object has - effectively anything that needs a datafetcher implementation.

This enables a workflow descirbed in https://github.com/Netflix/dgs-framework/discussions/234
